packages/components/src/sandbox/index.tsx

```diff
@@ -266,24 +177,35 @@ function SandBox( {
 		// See https://github.com/WordPress/gutenberg/pull/44378
 	}, [] );
 
-	useEffect( () => {
-		trySandBox();
-		// Passing `exhaustive-deps` will likely involve a more detailed refactor.
-		// See https://github.com/WordPress/gutenberg/pull/44378
-	}, [ title, styles, scripts ] );
-
-	useEffect( () => {
-		trySandBox( true );
-		// Passing `exhaustive-deps` will likely involve a more detailed refactor.
-		// See https://github.com/WordPress/gutenberg/pull/44378
-	}, [ html, type ] );
+	// Generate the HTML for srcdoc attribute
+	// Using srcdoc instead of document.write() preserves referrer information in Safari
+	const srcDocHtml = useMemo( () => {
+		const docLang =
+			typeof document !== 'undefined'
+				? document.documentElement.lang
+				: 'en';
+		return `<!DOCTYPE html>
+<html lang="${ docLang }" class="${ type || '' }">
+	<head>
+		<title>${ title }</title>
+		<style>${ style }</style>
+		${ styles.map( ( rules ) => `<style>${ rules }</style>` ).join( '' ) }
+	</head>
+	<body data-resizable-iframe-connected class="${ type || '' }">
+		<div>${ html }</div>
+		<script type="text/javascript">(${ observeAndResizeJS.toString() })();</script>
+		${ scripts.map( ( src ) => `<script src="${ src }"></script>` ).join( '' ) }
+	</body>
+</html>`;
+	}, [ html, title, type, styles, scripts ] );
 
 	return (
 		<iframe
 			ref={ useMergeRefs( [ ref, useFocusableIframe() ] ) }
 			title={ title }
 			tabIndex={ tabIndex }
 			className="components-sandbox"
+			srcDoc={ srcDocHtml }
```

Comment:
Using `srcdoc` rather than writing the HTML to the `iframe` ensures the referrer is preserved in Safari. 

These changes are more complex than the previous commit (https://github.com/WordPress/gutenberg/pull/73316/commits/6a07db1160913b9e4afe0196727594bccb6afcee) approach of simply using `WpEmbedPreview` for YouTube embed previews. These wider `Sandbox` changes may have repercussions for other `Sandbox` usages.